### ISSUE TITLE
flutter doctor crash: Cannot find executable for git

### ISSUE THREAD

**Issue by guna84** [flutter/flutter#57802]
**Update by @christopherfujino:** This was regressed in https://github.com/flutter/flutter/commit/06cd79d7dff4419956c9f2268e0d5a53a49e3fb2 (first published in `1.19.0-2.0.pre` and last published in `1.20.0-2.0.pre`) and will be fixed by https://github.com/flutter/flutter/pull/59369. It has been nominated for a cherry-pick into beta. This will only affect flutter users who do not have git installed.

To fix ideally you should install git for windows (it's listed as a dependency on the website https://flutter.dev/docs/get-started/install/windows), or alternatively you can add the path to mingit to your user path manually (it is located at `path\to\flutter\bin\mingit\cmd`).

---

```bash
Unhandled exception:
Invalid argument(s): Cannot find executable for git.
#0      _getExecutable (package:process/src/interface/local_process_manager.dart:127:5)
#1      LocalProcessManager.runSync (package:process/src/interface/local_process_manager.dart:94:30)
#2      _DefaultProcessUtils.runSync (package:flutter_tools/src/base/process.dart:417:51)
#3      _runGit (package:flutter_tools/src/version.dart:665:23)
#4      new FlutterVersion (package:flutter_tools/src/version.dart:54:26)
#5      runInContext.<anonymous closure> (package:flutter_tools/src/context_runner.dart:130:29)
#6      AppContext._generateIfNecessary.<anonymous closure> (package:flutter_tools/src/base/context.dart:104:41)
#7      _LinkedHashMapMixin.putIfAbsent (dart:collection-patch/compact_hash.dart:291:23)
#8      AppContext._generateIfNecessary (package:flutter_tools/src/base/context.dart:92:20)
#9      AppContext.get (package:flutter_tools/src/base/context.dart:121:32)
#10     flutterVersion (package:flutter_tools/src/globals.dart:88:46)
#11     new _DefaultUsage (package:flutter_tools/src/reporting/usage.dart:192:51)
#12     new Usage (package:flutter_tools/src/reporting/usage.dart:79:9)
#13     runInContext.<anonymous closure> (package:flutter_tools/src/context_runner.dart:187:20)
#14     AppContext._generateIfNecessary.<anonymous closure> (package:flutter_tools/src/base/context.dart:104:41)
#15     _LinkedHashMapMixin.putIfAbsent (dart:collection-patch/compact_hash.dart:291:23)
#16     AppContext._generateIfNecessary (package:flutter_tools/src/base/context.dart:92:20)
#17     AppContext.get (package:flutter_tools/src/base/context.dart:121:32)
#18     flutterUsage (package:flutter_tools/src/globals.dart:51:35)
#19     _handleToolError (package:flutter_tools/runner.dart:122:13)
#20     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:76:13)
#21     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:70:17)
#22     _rootRunBinary (dart:async/zone.dart:1222:13)
#23     _CustomZone.runBinary (dart:async/zone.dart:1107:19)
#24     runZonedGuarded.<anonymous closure> (dart:async/zone.dart:1601:18)
#25     _CustomZone.handleUncaughtError (dart:async/zone.dart:1076:19)
#26     Future._propagateToListeners (dart:async/future_impl.dart:646:16)
#27     Future._completeError (dart:async/future_impl.dart:537:5)
#28     _AsyncAwaitCompleter.completeError (dart:async-patch/async_patch.dart:47:15)
#29     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart)
#30     _asyncErrorWrapperHelper.errorCallback (dart:async-patch/async_patch.dart:91:64)
#31     _rootRunBinary (dart:async/zone.dart:1214:47)
#32     _CustomZone.runBinary (dart:async/zone.dart:1107:19)
#33     _FutureListener.handleError (dart:async/future_impl.dart:157:20)
#34     Future._propagateToListeners.handleError (dart:async/future_impl.dart:708:47)
#35     Future._propagateToListeners (dart:async/future_impl.dart:729:24)
#36     Future._completeError (dart:async/future_impl.dart:537:5)
#37     Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart:593:7)
#38     _rootRun (dart:async/zone.dart:1190:13)
#39     _CustomZone.run (dart:async/zone.dart:1093:19)
#40     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
#41     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
#42     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#43     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#44     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#45     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)
```

**Comment by iapicca** [comment]
Hi @guna84 
can you please run from your terminal
```console
flutter doctor -v
```
and provide the full logs
Thank you

**Comment by guna84** [comment]
![flutter doc](https://user-images.githubusercontent.com/65765838/82686319-4de13b80-9c73-11ea-984f-edd454321055.jpg)

**Comment by jmagman** [comment]
What version of Flutter do you have installed?

**Comment by DanielBaltazarSchneider** [comment]
It seems to be a problem with the latest version of the flutter. Delete your current version and install a previous version.

**Comment by christopherfujino** [comment]
@guna84 do you actually have git installed on your device? git is a requirement for flutter, as described here: https://flutter.dev/docs/get-started/install/windows

**Comment by Genderovic** [comment]
Hello friends, someone could tell me step by step the solution of this problem, I do not have so much knowledge in flutter and also I cannot find the GIT executable directory, most likely I do not have it installed. Thank.

**Comment by Daibaku9999** [comment]
I reinstalled older version and it fixed the problem.

**Comment by jmagman** [comment]
@Daibaku9999 You're saying you have git installed, hit this issue, and it was resolved by downgrading Flutter?  What version were you on, and what version did you downgrade to?

**Comment by AvelinoJavier** [comment]
I also have this problem. Before the last flutter update it worked perfect. I have git installed but it keeps showing the same message

[flutter] flutter doctor -v


Unhandled exception:
Invalid argument(s): Cannot find executable for git.
#0      _getExecutable (package:process/src/interface/local_process_manager.dart:127:5)
#1      LocalProcessManager.runSync (package:process/src/interface/local_process_manager.dart:94:30)
#2      _DefaultProcessUtils.runSync (package:flutter_tools/src/base/process.dart:417:51)
#3      _runGit (package:flutter_tools/src/version.dart:665:23)
#4      new FlutterVersion (package:flutter_tools/src/version.dart:54:26)
#5      runInContext.<anonymous closure> (package:flutter_tools/src/context_runner.dart:138:29)
#6      AppContext._generateIfNecessary.<anonymous closure> (package:flutter_tools/src/base/context.dart:104:41)
#7      _LinkedHashMapMixin.putIfAbsent (dart:collection-patch/compact_hash.dart:291:23)
#8      AppContext._generateIfNecessary (package:flutter_tools/src/base/context.dart:92:20)
#9      AppContext.get (package:flutter_tools/src/base/context.dart:121:32)
#10     flutterVersion (package:flutter_tools/src/globals.dart:88:46)
#11     new _DefaultUsage (package:flutter_tools/src/reporting/usage.dart:192:51)
#12     new Usage (package:flutter_tools/src/reporting/usage.dart:79:9)
#13     runInContext.<anonymous closure> (package:flutter_tools/src/context_runner.dart:195:20)
#14     AppContext._generateIfNecessary.<anonymous closure> (package:flutter_tools/src/base/context.dart:104:41)
#15     _LinkedHashMapMixin.putIfAbsent (dart:collection-patch/compact_hash.dart:291:23)
#16     AppContext._generateIfNecessary (package:flutter_tools/src/base/context.dart:92:20)
#17     AppContext.get (package:flutter_tools/src/base/context.dart:121:32)
#18     flutterUsage (package:flutter_tools/src/globals.dart:51:35)
#19     _handleToolError (package:flutter_tools/runner.dart:122:13)
#20     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:76:13)
#21     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:70:17)
#22     _rootRunBinary (dart:async/zone.dart:1222:13)
#23     _CustomZone.runBinary (dart:async/zone.dart:1107:19)
#24     runZonedGuarded.<anonymous closure> (dart:async/zone.dart:1601:18)
#25     _CustomZone.handleUncaughtError (dart:async/zone.dart:1076:19)
#26     Future._propagateToListeners (dart:async/future_impl.dart:646:16)
#27     Future._completeError (dart:async/future_impl.dart:537:5)
#28     _AsyncAwaitCompleter.completeError (dart:async-patch/async_patch.dart:47:15)
#29     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart)
#30     _asyncErrorWrapperHelper.errorCallback (dart:async-patch/async_patch.dart:91:64)
#31     _rootRunBinary (dart:async/zone.dart:1214:47)
#32     _CustomZone.runBinary (dart:async/zone.dart:1107:19)
#33     _FutureListener.handleError (dart:async/future_impl.dart:157:20)
#34     Future._propagateToListeners.handleError (dart:async/future_impl.dart:708:47)
#35     Future._propagateToListeners (dart:async/future_impl.dart:729:24)
#36     Future._completeError (dart:async/future_impl.dart:537:5)
#37     Future._asyncCompleteError.<anonymous closure> (dart:async/future_impl.dart:593:7)
#38     _rootRun (dart:async/zone.dart:1190:13)
#39     _CustomZone.run (dart:async/zone.dart:1093:19)
#40     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
#41     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
#42     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#43     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#44     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#45     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)
exit code 255

**Comment by jmagman** [comment]
@AvelinoJavier 
> What version were you on, and what version did you downgrade to?

**Comment by AvelinoJavier** [comment]
@jmagman 
flutter version
1.19.0-4.1.pre
1.19.0-5.0.pre
1.19.0-4.0.pre
1.17.3

I was testing directly from windows git bash and I was able to run flutter pub get successfully. Apparently the problem originated from wanting to run pub get automatically when saving pubspec.yaml from visual studio code.

**Comment by christopherfujino** [comment]
@AvelinoJavier so it works on `1.17.3`, but you get the exception on `1.19.0-4.0.pre` and after?

**Comment by AvelinoJavier** [comment]
@christopherfujino 
Yes, it worked with this version described below, running from the git terminal for windows. But it still doesn't work from visual studio code.

flutter --version
Flutter 1.19.0-4.1.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision f994b76974 (2 days ago) • 2020-06-09 15:53:13 -0700
Engine • revision 9a28c3bcf4
Tools • Dart 2.9.0 (build 2.9.0-14.1.beta)

**Comment by christopherfujino** [comment]
I'm taking a look at this now

**Comment by christopherfujino** [comment]
Hmm, I tested this on my windows machine, on both stable and beta, and I didn't get the error. Also, there is no diff on the file `packages/flutter_tools/lib/src/version.dart` from the stacktrace above.

@AvelinoJavier what do you mean by "it still doesn't work from visual studio code"? Do you mean the terminal in VS Code? I also do not get the exception running `flutter doctor -v` from the terminal in VSCode.

For reference, the command that the tool is calling when we hit this exception is:

```
git -c log.showSignature=false log -n 1 --pretty=format:%H
```

**Comment by christopherfujino** [comment]
I now suspect that it was this https://github.com/flutter/flutter/pull/57257/files#diff-962962d71e1e202f7b789b3bac77565cR31, where we refactored the flutter batch file entrypoint.

The following lines were moved from `flutter.bat` to `shared.bat`, I wonder if they're no longer working correctly? @goderbauer @jonahwilliams any idea?

```
REM If available, add location of bundled mingit to PATH
SET mingit_path=%FLUTTER_ROOT%\bin\mingit\cmd
IF EXIST "%mingit_path%" SET PATH=%PATH%;%mingit_path%
```

**Comment by jonahwilliams** [comment]
best way to handle this would be to pull this part back into flutter.bat and dart.bat, I'll open a PR

**Comment by christopherfujino** [comment]
> best way to handle this would be to pull this part back into flutter.bat and dart.bat, I'll open a PR

I'm locally testing this. If that was indeed the regression, I'll approve your PR @jonahwilliams

**Comment by christopherfujino** [comment]
Indeed, I can confirm that #59369 will fix this issue.

**Comment by christopherfujino** [comment]
> Hello friends, someone could tell me step by step the solution of this problem, I do not have so much knowledge in flutter and also I cannot find the GIT executable directory, most likely I do not have it installed. Thank.

If you install git for windows via https://git-scm.com/download/win then flutter should work.

**Comment by AvelinoJavier** [comment]
Installing a new version of git solved my problem. Thank you

**Comment by deakjahn** [comment]
"you should install git for windows"

Considering the Flutter comes with its own copy, couldn't this be reconsidered? I know it's not that much relevant in this day and age but, simply, it feels just very wrong that you need a nearly 1 GB software just to use a few command line executables. Yes, I know that it could be explained away why Git for Windows it so large, bringing along everything and the kitchen sink with MinGW but that doesn't make it any less ridiculous.

**Comment by deakjahn** [comment]
@christopherfujino "git is a requirement for flutter"

I don't know since when but I definitely never needed it *before* I actually started to maintain a project and really needed it to use the integrated Git support in Android Studio. For about a year when I only used packages or wrote my local ones but haven't published any, Git was never required (Flutter could have used its own bundled one, I don't know, but I never had to install it for sure).

**Comment by deakjahn** [comment]
I'm happy to report that it **does work** with just the env variables pointing to the bundled MinGit. So please, don't suggest to install Git for Windows as a workaround. It isn't needed.

**Comment by christopherfujino** [comment]
> @christopherfujino "git is a requirement for flutter"
> 
> I don't know since when but I definitely never needed it _before_ I actually started to maintain a project on GitHub and actually needed it to use the integrated Git support in Android Studio. For about a year of Flutter development while I only used packages or wrote my local ones but haven't published any, Git was never required (Flutter could have used its own bundled one behind my back, I don't know, but I never had to install it for sure).

@deakjahn Git has always been officially listed as a system requirement. It happened to work without it because mingit was vendored with it to reduce Flutter onboarding friction. In the case of this issue, which has already been fixed upstream, users whose current flutter checkout has the regression and who do not have git installed will require manual intervention to allow them to `flutter upgrade` to get the fix. If manual intervention is required anyway and they are missing a required dependency, I think the best recommendation for users is to install git.

**Comment by deakjahn** [comment]
Yes, I'm one of those users. Quite accidentally, a broken video card forced a complete system reinstall for me just at the same time this regression arrived. This meant that although I had Git for Win previously installed, now it caught me when Android Studio was already up and running. And much to my relief, as I installed, I saw that MinGit is trying to go out through the firewall. So, I intentionally stayed away from Git for Win this time and fixed my .bat files with that fix manually. And I also set up Android Studio to use the bundled MinGit and it works flawlessly. As long as this remains so, I see no need for Git for Win.
